File: [0358][Hard][Rearrange_String_k_Distance_Apart]/Rearrange_String_k_Distance_Apart_2.java
package java.util;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.Comparators;
// solution 2, max heap.
// leetcode time     cost : 49 ms
// leetcode memory   cost : 45.7 MB 
// Time  Complexity: O(time)
// Space Complexity: O(1)
class Solution {
    public String rearrangeString(String s, int k) {
        if(k <= 1)
            return s;
        int[] count = new int[26];
        for(char c: s.toCharArray())
            ++count[c - 'a'];
        PriorityQueue<Integer> priorityQueue = new PriorityQueue<>(new Comparator<Integer>() {
            public int compare(Integer o1, Integer o2) {
                return count[o2] - count[o1] == 0 ? o1 - o2 : count[o2] - count[o1];
            }
        });
        
        for(int i = 0; i < 26; ++i)
            if(count[i] != 0)
            priorityQueue.add(i);
        StringBuilder strBuilder = new StringBuilder();
        for(int i = 0; i < s.length();  i += k) {
            List<Integer> temp = new ArrayList<>();
            for(int j = 0; j < k && i + j < s.length(); ++j) {
                if(priorityQueue.isEmpty())
                    return "";
                int c = priorityQueue.poll();
                temp.add(c);
                --count[c];
                strBuilder.append((char)('a' + c));
            }
            for(int c: temp)
                if(count[c] != 0)
                priorityQueue.add(c); 
        }       
        return strBuilder.toString();
    }
}

File: [0358][Hard][Rearrange_String_k_Distance_Apart]/Rearrange_String_k_Distance_Apart_2.py
# solution 2, solution 3, analysis and design..
# leetcode time     cost : 116 ms
# leetcode memory   cost : 13.8 MB 
# Time  Complexity: O(M)
# Space Complexity: O(1)
from collections import Counter
import collections
class Solution:
    def rearrangeString(self, s: str, k: int) -> str:
        if k <= 1: return s
        c = Counter(s)
        n = len(s)
        #set count value as negative value to match the min heap usage, the min top is the max counter.
        charPair = []
        for key, v in c.items():
            charPair = charPair + [[key,v]]            
        charDictList = sorted(charPair,key=lambda x:x[1],reverse=False)
        result = ""
        maxIndex = len(charDictList)-1
        while charDictList[maxIndex][1]:
            i = 0;
            validCharInSubLoop =0;
            while (i < k):
                if (charDictList[maxIndex][1] == 0):
                    break;
                if (i <= maxIndex and charDictList[maxIndex-i][1] > 0):
                    result = result + charDictList[maxIndex-i][0]
                    charDictList[maxIndex-i][1] = charDictList[maxIndex-i][1] -1
                    #used to counter the valid char count in this sub group
                    validCharInSubLoop= validCharInSubLoop +1
                i = i + 1
                # for sub loop not include last loop, if sub loop finished with no enough valid char can be use.
                if (charDictList[maxIndex][1] > 0 and i == k and validCharInSubLoop != k): 
                    return ""
            charDictList = sorted(charDictList,key=lambda x:x[1],reverse=False)
        return result 
    
def main():
    str1 = "aabbccdddfffff"         #expect is "fdafbcfdafbcdf"
    str2 = "aabbccdddffffff";         #expect is ""
    k = 3
    s = Solution()
    res = s.rearrangeString(str1,k) 
    print("in python,rearrangeString res is : "+res)
    res = s.rearrangeString(str2,k) 
    print("in python,rearrangeString res is : "+res)
if __name__ =='__main__':
    main()    